dss/api/collections.py

```diff
@@ -0,0 +1,190 @@
+    if not idempotent:
+        raise DSSException(requests.codes.conflict,
+                           f"collection_tombstone_already_exists",
+                           f"collection tombstone with UUID {uuid} already exists")
+    status_code = requests.codes.ok
+    response_body = dict()  # type: dict
+
+    return jsonify(response_body), status_code
+
+@functools.lru_cache(maxsize=64)
+def get_json_metadata(entity_type: str, uuid: str, version: str, replica: Replica, blobstore_handle: BlobStore):
+    try:
+        key = "{}s/{}.{}".format(entity_type, uuid, version)
+        # TODO: verify that file is a metadata file
+        size = blobstore_handle.get_size(replica.bucket, key)
+        if size > MAX_METADATA_SIZE:
+            raise DSSException(422, "invalid_link",
+                               "The file UUID {} refers to a file that is too large to process".format(uuid))
+        return json.loads(blobstore_handle.get(
+            replica.bucket,
+            "{}s/{}.{}".format(entity_type, uuid, version)))
+    except BlobNotFoundError as ex:
+        raise DSSException(404, "invalid_link", "Could not find file for UUID {}".format(uuid))
+
+def resolve_content_item(replica: Replica, blobstore_handle: BlobStore, item: dict):
+    try:
+        if item["type"] in {"file", "bundle", "collection"}:
+            item_metadata = get_json_metadata(item["type"], item["uuid"], item["version"], replica, blobstore_handle)
+        else:
+            item_metadata = get_json_metadata("file", item["uuid"], item["version"], replica, blobstore_handle)
+            if "fragment" not in item:
+                raise Exception('The "fragment" field is required in collection elements '
+                                'other than files, bundles, and collections')
+            blob_path = "blobs/" + ".".join((
+                item_metadata[FileMetadata.SHA256],
+                item_metadata[FileMetadata.SHA1],
+                item_metadata[FileMetadata.S3_ETAG],
+                item_metadata[FileMetadata.CRC32C],
+            ))
+            # check that item is marked as metadata, is json, and is less than max size
+            item_doc = json.loads(blobstore_handle.get(replica.bucket, blob_path))
+            item_content = jsonpointer.resolve_pointer(item_doc, item["fragment"])
+            return item_content
+    except DSSException:
+        raise
+    except Exception as e:
+        raise DSSException(
+            422,
+            "invalid_link",
+            'Error while parsing the link "{}": {}: {}'.format(item, type(e).__name__, e)
+        )
+
+def verify_collection(contents: List[dict], replica: Replica, blobstore_handle: BlobStore):
+    """
+    Given user-supplied collection contents that pass schema validation, resolve all entities in the collection and
+    verify they exist.
+    """
+    executor = ThreadPoolExecutor(max_workers=8)
```

Comment:
Request: Use TPE as a context manager to ensure timely shutdown.

Reply:
That's the wrong thing to do. Using the ThreadPoolExecutor context manager will cause it to wait for all futures before leaving the frame. We deliberately want to return upon the first error before leaving the frame. When we do so, the futures that were running at that point will finish, but no new futures will run.

Reply:
> When we do so, the futures that were running at that point will finish, but no new futures will run.

I don't think so:

```py
from concurrent.futures import ThreadPoolExecutor
tpe = ThreadPoolExecutor(8)

def fail(x):
    assert False

l = list(range(100000))
for x in tpe.map(fail, l):
    pass
```

Note how it blocks for a long time. If you don't notice the blocking, add a zero to the end of the range.

Even if we changed this to using `tpe.submit()` and managed to bail out early, we'd still have to demonstrate that all worker threads are automatically shut down right away. Otherwise, we'd risk accumulating large amounts of threads over a short period of time if multiple PUT or PATCH requests come in.

So while I agree that bailing out early is the scalable thing to do, I'd be really surprised if fire and forget is the right approach here. That's why I suggested batching below: you have an upper bound on the number of failures but you do report multiple failures at once, which is user-friendly (and, assuming they fix all the issues before retrying the request, resource-friendly.

Reply:
You are right. What I said is true for long-running futures that block on I/O (which ours are), but can't be said in general. I think my original approach is still safe and efficient given the specifics of where and how our threads spend our time, but I'm fine with the batch+wait approach, since it's barely any more complex.

As an immediate solution, I've implemented batching and the context manager.